File: Data Structures & Algorithms/valid-palindrome-ii/submission-2.cpp
class Solution {
public:
    bool validPalindrome(string s) {
        int l =0 , r = s.length() -1 ;
        while (l<r){
            if(s[l] == s[r]){
                l++;
                r--;
            }
            else{
                return ispal(s,l+1 , r) ||
                       ispal(s ,l ,r-1);
            }
        }
        
    }

private:
    bool ispal(string s , int l , int r){
        while (l < r){
            if (s[l] != s[r]){
                return false;
            }
            l++;
            r--;
        }
        return true;
    }
};

File: Data Structures & Algorithms/palindrome-partitioning/submission-2.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        res = []
        pal = []

        def backtrack(start):
            if start == len(s):
                res.append(pal.copy())
                return

            for end in range(start, len(s)):
                if self.isPal(s, start, end):
                    pal.append(s[start:end+1])
                    backtrack(end+1)
                    pal.pop()

        backtrack(0)
        return res

    def isPal(self, s, l, r):
        while l < r:
            if s[l] != s[r]:
                return False
            l, r = l+1, r-1
        return True


